### PR TITLE
Added max result size limit for IMap.values(), IMap.keySet() and IMap.entrySet()

### PR DESCRIPTION
**Pull request by Donnerbart**
Added max result size limit for IMap.values(), IMap.keySet() and IMap.entrySet(). Will also limit the result size of Query operations.

### CODE REVIEW

**Comment by Donnerbart** [comment]
@pveentjer \*poke\*

**Comment by Donnerbart** [comment]
verify

**Comment by Donnerbart** [comment]
@pveentjer \*poke\*

**Comment by pveentjer** [comment]
Finally time.. *making some tea and then getting started*

File: hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java [hunk 1]
```diff
@@ -214,6 +214,17 @@
      */
     public static final String PROP_JCACHE_PROVIDER_TYPE = "hazelcast.jcache.provider.type";
 
+    /**
+     * Result size limit of a predicate query operation.
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java]: Can you provide a bit more information? E.g is this in total? Per member? Per partition? And why do we want this? We don't we get all results in one go? This is going to be asked by people. 

Also explain how it can be disabled so that HZ is working in old (but potentially dangerous) mode.

File: hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java [hunk 2]
```diff
@@ -214,6 +214,17 @@
...
+     */
+    public static final String PROP_QUERY_RESULT_SIZE_LIMIT = "hazelcast.query.result.size.limit";
+
+    /**
+     * Maximum value of local partitions to trigger local pre-check for <tt>TruePredicate</tt> query operations on maps.
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java]: Can you provide a bit more information for the mortals like me?

File: hazelcast/src/test/java/com/hazelcast/map/MapUnboundReturnValues_StringKeyTest.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+import com.hazelcast.test.HazelcastParallelClassRunner;
+import com.hazelcast.test.annotation.SlowTest;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category(SlowTest.class)
+public class MapUnboundReturnValues_StringKeyTest extends MapUnboundReturnValuesTestSupport {
```
Comment (pveentjer) [inline hazelcast/src/test/java/com/hazelcast/map/MapUnboundReturnValues_StringKeyTest.java]: Can't we create a few fast tests? It is good to have nightly tests, but I would like to get some coverage during the day :)

File: hazelcast/src/main/java/com/hazelcast/map/impl/DataLimitExceededException.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package com.hazelcast.map.impl;
+
+import com.hazelcast.core.HazelcastException;
+
+public class DataLimitExceededException
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/map/impl/DataLimitExceededException.java]: Isn't it query result size exceeded? Data limit exceeded makes my think of my mobile internet provider.

File: hazelcast/src/test/java/com/hazelcast/map/MapUnboundReturnValues_StringKeyTest.java (hunk 3 above)
Comment (Donnerbart) [inline hazelcast/src/test/java/com/hazelcast/map/MapUnboundReturnValues_StringKeyTest.java]: Hmm, that could be hard, since the feature depends heavily on the distribution of data in the cluster. This is why there is a hard-coded minimum of 100k before the feature triggers at all.

What makes the tests so slow is that I check that the feature doesn't trigger below the configured limit.

If we drop that check we could fill a single node with 120k small elements and see if the exception is thrown:
* filling cluster up to a lower bound which should be safe to not trigger
* call `IMap.values()` to check that lower bound didn't trigger
* repeat:
  * fill a bit more data in
  * call `IMap.values()`

If we drop this check we could just fill a single node with 120k items and run a single `Imap.values()` on it to trigger the exception. This should be much faster (single node, just one query).

How long may that test run? I'll give it a try locally.

Comment (Donnerbart) [inline hazelcast/src/test/java/com/hazelcast/map/MapUnboundReturnValues_StringKeyTest.java]: BTW: `MapUnboundReturnValues_PreCheckTest.testMapKeySet_SmallLimit()` takes 16 to 17 seconds on my local machine. I think that is a `SlowTest`.

Comment (Donnerbart) [inline hazelcast/src/test/java/com/hazelcast/map/MapUnboundReturnValues_StringKeyTest.java]: With the speedup (and less checks) I could provide a test with ~10 seconds runtime.

File: hazelcast/src/main/java/com/hazelcast/map/impl/BasicMapContextQuerySupport.java [hunk 5]
```diff
@@ -42,13 +44,17 @@
 
     private final MapServiceContext mapServiceContext;
     private final NodeEngine nodeEngine;
+    private final long maxLocalPartitionsLimitForPreCheck;
 
     public BasicMapContextQuerySupport(MapServiceContext mapServiceContext) {
         this.mapServiceContext = mapServiceContext;
         this.nodeEngine = mapServiceContext.getNodeEngine();
+        this.maxLocalPartitionsLimitForPreCheck = Math.abs(
+                nodeEngine.getGroupProperties().QUERY_MAX_LOCAL_PARTITION_LIMIT_FOR_PRE_CHECK.getInteger());
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/BasicMapContextQuerySupport.java]: Do we need to handle special case for `Integer.MIN_VALUE` when using `Math.abs` ?

Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/map/impl/BasicMapContextQuerySupport.java]: That's a good catch! Do we have a common place where I can prepare a `GroupProperty` once, so I don't have to do those checks on every query?

**Comment by jerrinot** [comment]
@Donnerbart: Have you tried to measure impact of on performance?

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/QueryOperation.java [hunk 6]
```diff
@@ -115,53 +123,56 @@ public void run() throws Exception {
         checkPartitionStateChanges(partitionService, partitionStateVersion);
     }
 
-    private void checkPartitionStateChanges(InternalPartitionService partitionService, int partitionStateVersion) {
-        if (partitionStateVersion != partitionService.getPartitionStateVersion()) {
-            getLogger().info("Partition assignments changed while executing query: " + predicate);
+    public static int getMaxResultLimit(NodeEngine nodeEngine) {
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/QueryOperation.java]: it would be nice to have some JavaDoc here. moreover - should this be really a responsibility of QueryOperation?

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/QueryOperation.java [hunk 7]
```diff
@@ -115,53 +123,56 @@ public void run() throws Exception {
...
+        int maxResultLimit = Math.abs(nodeEngine.getGroupProperties().QUERY_RESULT_SIZE_LIMIT.getInteger());
+        if (maxResultLimit > 0 && maxResultLimit < MINIMUM_MAX_RESULT_LIMIT) {
+            maxResultLimit = MINIMUM_MAX_RESULT_LIMIT;
+            nodeEngine.getLogger(QueryOperation.class)
+                      .warning("Max result limit was set to minimal value of " + MINIMUM_MAX_RESULT_LIMIT);
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/QueryOperation.java]: this will be written to a log for each query?

File: hazelcast/src/main/java/com/hazelcast/map/impl/QueryResultSizeExceededException.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package com.hazelcast.map.impl;
+
+import com.hazelcast.core.HazelcastException;
+
+public class QueryResultSizeExceededException
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/map/impl/QueryResultSizeExceededException.java]: a public Exception should definitely have a good JavaDoc

File: hazelcast/src/main/java/com/hazelcast/map/impl/BasicMapContextQuerySupport.java (hunk 5 above)
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/BasicMapContextQuerySupport.java]: afaik no we don't

File: hazelcast/src/main/java/com/hazelcast/map/impl/QueryResultSizeExceededException.java (hunk 8 above)
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/QueryResultSizeExceededException.java]: also it should not be under `impl` package i think

**Comment by Donnerbart** [comment]
Thank you all for your valuable reviews!

* I encapsulated most of the functionality in the new class `MapQueryResultSizeLimitHelper`, which is used by `BasicMapContextQuerySupport` and `QueryOperation` now
* the `GroupProperties` are just evaluated once when the map context is created
* the `QueryResultSizeExceededException` is now in the package `com.hazelcast.map`, I hope that is a better place
* there is more JavaDoc, especially in the `GroupProperties`

**Comment by Donnerbart** [comment]
@jerrinot No performance test yet, but I hope the impact is minimal in the way I implemented the feature.  The normal result size checks should be very cheap.

The highest risk will be increased latency due to the local pre-check, which is a really new operation within the performance path. But it very easy to disable or maybe should even be disabled by default.

**Comment by Donnerbart** [comment]
verify

File: hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java [hunk 9]
```diff
@@ -221,6 +227,37 @@
      */
     public static final String PROP_JCACHE_PROVIDER_TYPE = "hazelcast.jcache.provider.type";
 
+    /**
+     * Result size limit for query operations on maps.
+     * <p/>
+     * This value defines the maximum number of returned elements for a single query result. If a query exceeds this number of
+     * elements a {@link QueryResultSizeExceededException} will be thrown.
+     * <p/>
+     * This feature is in place to prevent an OOME if a single node is requesting the whole data set of the cluster, e.g. by
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java]: it applies if and only if the TruePredicate is used, doesn't it?

Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java]: Just the local pre-check should be limited to the usage of `TruePredicate` (directly or indirectly). The query result limit (checked in `QueryOperation`) should work for all queries.

Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java]: The local pre-check doesn't work if the query won't return all values (it's just summing up the partition sizes). So we need that check for `TruePredicate` there. The check in `QueryOperation` checks the real query results, so it will also work for all types of queries.

**Comment by Donnerbart** [comment]
I just made run in the test lab with a map below the result size limit (so all `IMap.values()` were executed without throwing any exception). I compared the actual 3.5 master with the feature branch. I can't see a big difference in the latency:

Imgur

**Comment by Donnerbart** [comment]
I could just get rid of a bit of runtime calculations in `MapQueryResultSizeLimitHelper.getNodeResultLimit()` since most of the values are runtime constant. I also removed the partition count as parameter.

**Comment by Donnerbart** [comment]
Added new test class for the encapsulated helper class. Code coverage is now 100%. Found an of by one error in the group properties parsing with this.

**Comment by Donnerbart** [comment]
verify

**Comment by Donnerbart** [comment]
verify

**Comment by jerrinot** [comment]
I like the current code, I think it can be merged. However I don't understand the latency histogram. Values on the Y-axis looks off. There should be a number of operations for a given latency (as on the X-axis). Hence it always has to be > 1.

**Comment by Donnerbart** [comment]
verify

**Comment by Donnerbart** [comment]
You are right, that looks suspicious. I will have a look with an older Simulator version, before I touched anything in the probes and visualizer. Maybe I messed up something in the time measurement.

Beside that there is still the objection of Peter with the `0` vs. `Integer.MAX_VALUE` to disable the feature (see TDD comments in the wiki). Any opinion about this? It would be quite easy to change, I just don't like the idea to have a special value on two sides of the limit range (<= 0 would be invalid, Integer.MAX_VALUE would be disabled).

**Comment by Donnerbart** [comment]
Updated the feature, so `-1` disabled both configurable values. There are unit tests for all edge cases. The documentation in `GroupProperties` is updated.

**Comment by Donnerbart** [comment]
Rebased, local checkstyle/findbugs was happy, all new tests are still green :)

**Comment by pveentjer** [comment]
I see that we have a different latency distribution graph. I hope the original is still available.

**Comment by Donnerbart** [comment]
What do you mean with this? Do you want to see the probe XML files?

**Comment by pveentjer** [comment]
Normally we get something like this:

http://www.fhwa.dot.gov/publications/research/safety/09049/images/image031.jpg

**Comment by Donnerbart** [comment]
I'm sure this is some kind of bug, introduced by the probes refactoring or the visualizer refactoring. Not a stone was left standing there. To check this is the last Simulator framework thing on my agenda.

**Comment by Donnerbart** [comment]
For the interpretation: I think the scale may be broken, but the relation between both measurings should still be okay. So there was not big shift in latency with my feature. Nevertheless it's hard to read in the current state.

**Comment by jerrinot** [comment]
:+1: 
Good work!

File: hazelcast/src/main/java/com/hazelcast/map/impl/BasicMapContextQuerySupport.java [hunk 10]
```diff
@@ -288,16 +316,19 @@ private Future queryOnLocalMember(String mapName, Predicate predicate, NodeEngin
...
             throws ExecutionException, InterruptedException {
-        for (Future future : futures) {
-            QueryResult queryResult = getQueryResult(future);
+        for (Future<QueryResult> future : futures) {
+            QueryResult queryResult = future.get();
             if (queryResult == null) {
                 continue;
             }
+            if (queryResult.isResultLimitExceeded()) {
+                throw new QueryResultSizeExceededException();
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/map/impl/BasicMapContextQuerySupport.java]: you could add an explanation what's going on.

File: hazelcast/src/main/java/com/hazelcast/map/impl/MapQueryResultSizeLimitHelper.java [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+    public long getNodeResultLimit(int ownedPartitions) {
+        return (long) Math.ceil(resultLimitPerPartition * ownedPartitions);
+    }
+
+    boolean isPreCheckEnabled() {
+        return isPreCheckEnabled;
+    }
+
+    void checkMaxResultLimitOnLocalPartitions(String mapName) {
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/map/impl/MapQueryResultSizeLimitHelper.java]: could you split this method?

Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/map/impl/MapQueryResultSizeLimitHelper.java]: I don't know a good design pattern to split up those "if (check) return; if (check) return;" methods. I extracted one method, but I don't see any obvious possibilities to refactor something else.

**Comment by Donnerbart** [comment]
I added some JavaDoc to the `IMap` interface to make the users aware of the new feature and the consequences.

**Comment by Donnerbart** [comment]
verify

**Comment by bilalyasar** [comment]
verify

**Comment by bilalyasar** [comment]
verify

**Comment by bilalyasar** [comment]
verify

**Comment by Donnerbart** [comment]
Latency distribution with the actual feature branch vs. actual master:
Imgur

File: hazelcast/src/main/java/com/hazelcast/map/impl/MapQueryResultSizeLimitHelper.java (hunk 11 above)
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/map/impl/MapQueryResultSizeLimitHelper.java]: I guess it's fine now. it's just a smell when a see a long(-ish) method structured like
```
myMethod() {
//do A
stuff();
stuff();
stuff();

//do B
stuff();
stuff();

//do C
stuff();
stuff();
}
```

but it's alright now.

**Comment by jerrinot** [comment]
I see you added some warnings/comments when the `QueryResultSizeExceededException` is thrown. 

What I meant was more like a message in the exception itself - informing a user what's the current settings and possibly even how to change it.

**Comment by Donnerbart** [comment]
Here are both latency distribution graphs (actual feature branch vs. actual master).

Map filled closely below trigger limit (100k * 1.15 * 0.5 = 103500 elements):
Imgur

Almost empty map (I think there were 4 elements in the map, due to the warmup algorithm):
Imgur

**Comment by Donnerbart** [comment]
verify

**Comment by Donnerbart** [comment]
I updated the JavaDoc in `IMap` to make clear that the exception is just thrown on the server side.

I also added a test for the client side to check that the exception is not thrown.

**Comment by Donnerbart** [comment]
Good that I wrote that client test. Of course the feature kicked in when the client uses predicates, but no exception was thrown. So the result was silently cut. The same was happening for transactional maps.

I wrote tests for the client which check that the normal `IMap.values()`, `IMap.keySet()` and `IMap.entrySet()` throw no exception and return the full result size. I also wrote tests to check that the predicate methods throw the exception.

I still have to write tests for the transactional map. Also the normal, unchanged `IMap.entrySet()` on client side seems to run forever. Maybe there is a big with large data sets in that method.

**Comment by Donnerbart** [comment]
My tests are slow, so I put them in the nightly category. So they are not executed by the PR builder.

There are still failures though, so please do not merge yet.

**Comment by Donnerbart** [comment]
I will split up this PR, since it is already to big and more changes are coming (at least more test code).

Thank you all for the reviews and the comments, nothing is lost just moved to separate commits :)
